src/key.h

```diff
@@ -235,6 +242,71 @@ struct CExtKey {
     void SetSeed(Span<const std::byte> seed);
 };
 
+/** KeyPair
+ *
+ *  Wraps a `secp256k1_keypair` type, an opaque data structure for holding a secret and public key.
+ *  This is intended for BIP340 keys and allows us to easily determine if the secret key needs to
+ *  be negated by checking the parity of the public key. This class primarily intended for passing
+ *  secret keys to libsecp256k1 functions expecting a `secp256k1_keypair`. For all other cases,
+ *  CKey should be preferred.
+ */
+class KeyPair
+{
+public:
+    KeyPair() noexcept = default;
+    KeyPair(KeyPair&&) noexcept = default;
+    KeyPair& operator=(KeyPair&&) noexcept = default;
+    KeyPair& operator=(const KeyPair& other)
+    {
+        if (this != &other) {
+            if (other.m_keypair) {
+                MakeKeyPairData();
+                *m_keypair = *other.m_keypair;
+            } else {
+                ClearKeyPairData();
+            }
+        }
+        return *this;
+    }
+
+    KeyPair(const KeyPair& other) { *this = other; }
+
+    friend KeyPair CKey::ComputeKeyPair() const;
+    [[nodiscard]] bool SignSchnorr(const uint256& hash, Span<unsigned char> sig, const uint256& aux) const;
+
+    /**
+     * data() is provided as a read-only method for passing a KeyPair object to secp256k1 functions
+     * expecting a `secp256k1_keypair`. This avoids needing to create a temporary `secp256k1_keypair`
+     * object by allowing the KeyPair to be passed directly in the following manner:
+     *
+     *     reinterpret_cast<const secp256k1_keypair*>(keypair.data())
+     *
+     * Recall that `secp256k1_keypair` is an opaque data type, so this method should only be used
+     * for passing a KeyPair object as a secp256k1_keypair and should never be used to access the
+     * underlying keypair bytes directly.
+     */
+    const unsigned char* data() const { return IsValid() ? m_keypair->data() : nullptr; }
```

Comment:
If there is not a better alternative, I would suggest `secp256k1_keypair()` as a candidate for renaming this `data()` member function to further signify its use through the cast and be called like so:

`reinterpret_cast<const secp256k1_keypair*>(keypair.secp256k1_keypair())`

Reply:
(just realised I need to update the description of this PR to be a bit more clear on the motivation)

@sipa - my main motivations for having the `KeyPair` class is to a) encapsulate the taptweak logic so that it can be used outside of signing and b) to be able to pass the `KeyPair` directly to `secp256k1_*` functions expecting a `const secp256k1_keypair *`. For the latter, I think we will need some sort of public function. Here is an example of how this used in #28201:

https://github.com/bitcoin/bitcoin/blob/38067a6ef07d9b5108e75207960bdb7a21c14b70/src/common/bip352.cpp#L206-L221

I do agree a rename would be much better here, perhaps `ReadKeyPairData()/GetKeyPairData()`, or something like @itornaza 's suggestion, `secp256k1_keypair()`?.

Another option is leave the read-only public method out of this PR (since it's not currently used) and add the read-only method in #28201 ?

Reply:
I've removed `data()` for now and will add `GetKeyPairData` as a read-only method in #28201